CONTRIBUTING.md

```diff
@@ -0,0 +1,84 @@
+      ```
+
+   2. Using docker:
+
+      ```
+      ./ci/run_envoy_docker.sh './ci/do_ci.sh bazel.test'
+      ./ci/run_envoy_docker.sh './ci/do_ci.sh bazel.docs'
+      ```
+3. All configuration changes should have temporary associated documentation. Fields should be
+   hidden from the documentation via the `[#not-implemented-hide:]` comment tag. E.g.,
+
+   ```
+   // [#not-implemented-hide:] Some new cool field that I'm going to implement and then
+   // come back and doc for real!
+   string foo_field = 3;
+   ```
+
+   Additionally, [constraints](https://github.com/lyft/protoc-gen-validate/blob/master/README.md)
+   should be specified for new fields if applicable. E.g.,
+
+   ```
+   string endpoint = 2 [(validate.rules).message.required = true];
+   ```
+
+4. Next, the feature should be implemented in Envoy. New versions of data-plane-api are brought
+   in via editing [this](https://github.com/envoyproxy/envoy/blob/master/bazel/repository_locations.bzl)
+   file.
+5. Once (4) is completed, come back here and unhide the field from documentation and complete all
+   documentation around the new feature. This may include architecture docs, etc. Optimally, the
+   PR for documentation should be reviewed at the same time that the feature PR is reviewed in
+   the Envoy repository. See the following section for tips on writing documentation.
+
+## Documentation changes
+
+The Envoy project takes documentation seriously. We view it as one of the reasons the project has
+seen rapid adoption. As such, it is required that all features have complete documentation. This is
+generally going to be a combination of API documentation as well as architecture/overview
+documentation. The documentation can be built locally in the root of this repo via:
+
+```
+docs/build.sh
+```
+
+Or to use a hermetic docker container:
+
+```
+./ci/run_envoy_docker.sh './ci/do_ci.sh bazel.docs'
+```
+
+This process builds RST documentation directly from the proto files, merges it with the static RST
+files, and then runs [Sphinx](http://www.sphinx-doc.org/en/stable/rest.html) over the entire tree to
+produce the final documentation. The following are some general guidelines around documentation.
+
+* Cross link as much as possible. Sphinx is fantastic at this. Use it! See ample examples with the
+  existing documentation as a guide.
+* Please use a **single space** after a period in documentation so that all generated text is
+  consistent.
+* Comments can be left inside comments if needed (that's pretty deep, right?) via the `[#comment:]`
+  special tag. E.g.,
+
+  ```
+  // This is a really cool field!
+  // [#comment:TODO(mattklein123): Do something cooler]
+  string foo_field = 3;
+  ```
+
+* Prefer *italics* for emphasis as `backtick` emphasis is somewhat jarring in our Sphinx theme.
+* All documentation is expected to use proper English grammar with proper punctuation. If you are
+  not a fluent English speaker please let us know and we will help out.
```

Comment:
Could we split this piece and put it a separate markdown? PROTOC-DOC-HOW-TO or something? We have a bunch of thingies in the proto.. like top level comment, fields to hide, and more to come. Perhaps point people to a wiki? which would be more upto date and easier to maintain.